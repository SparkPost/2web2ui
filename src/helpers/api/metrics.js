const METRICS_BASE_URL = `/v1/metrics`;
const DELIVERABILITY_BASE_URL = `${METRICS_BASE_URL}/deliverability`;
const BENCHMARK_BASE_URL = `${METRICS_BASE_URL}/benchmarks`;

export function getDomains(params) {
  return {
    method: 'GET',
    url: `${METRICS_BASE_URL}/domains`,
    params,
  };
}

export function getCampaigns(params) {
  return {
    method: 'GET',
    url: `${METRICS_BASE_URL}/campaigns`,
    params,
  };
}

export function getSubjectCampaigns(params) {
  return {
    method: 'GET',
    url: `${METRICS_BASE_URL}/subject-campaigns`,
    params,
  };
}

export function getSendingIps(params) {
  return {
    method: 'GET',
    url: `${METRICS_BASE_URL}/sending-ips`,
    params,
  };
}

export function getIpPools(params) {
  return {
    method: 'GET',
    url: `${METRICS_BASE_URL}/ip-pools`,
    params,
  };
}

export function getTemplates(params) {
  return {
    method: 'GET',
    url: `${METRICS_BASE_URL}/templates`,
    params,
  };
}

export function getDeliverability(params, path) {
  const joinedPath = `${DELIVERABILITY_BASE_URL}${path ? `/${path}` : ''}`;
  return {
    method: 'GET',
    url: joinedPath,
    params,
  };
}

export function getTimeSeries(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/time-series`,
    params,
  };
}

export function getBounceClassification(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/bounce-classification`,
    params,
  };
}

export function getBounceReason(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/bounce-reason`,
    params,
  };
}

export function getBounceReasonByDomain(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/bounce-reason/domain`,
    params,
  };
}

export function getRejectionReasonByDomain(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/rejection-reason/domain`,
    params,
  };
}

export function getDelayReasonByDomain(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/delay-reason/domain`,
    params,
  };
}

export function getAttempted(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/deliverability/attempt`,
    params,
  };
}

export function getEngagement(params) {
  return {
    method: 'GET',
    url: `${DELIVERABILITY_BASE_URL}/link-name`,
    params,
  };
}

export function getIndustryBenchmark(params) {
  return {
    method: 'GET',
    url: `${BENCHMARK_BASE_URL}/inbox-rate`,
    params,
  };
}
