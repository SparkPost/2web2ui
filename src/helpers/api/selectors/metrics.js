export const selectRecipientDomains = ({ domains = [] } = {}) =>
  domains.map(value => ({
    type: 'domains',
    value,
  }));

export const selectSubaccounts = (subaccounts = []) =>
  subaccounts.map(({ name, id }) => ({
    type: 'subaccounts',
    value: `${name} (ID ${id})`,
    id,
  }));

export const selectSendingDomains = (sendingDomains = []) =>
  sendingDomains.map(({ domain }) => ({
    type: 'sending_domains',
    value: domain,
  }));

export const selectCampaigns = ({ campaigns = [] } = {}) =>
  campaigns.map(value => ({
    type: 'campaigns',
    value,
  }));

export const selectSubjectCampaigns = ({ 'subject-campaigns': subjectCampaigns = [] } = {}) =>
  subjectCampaigns.map(value => ({
    type: 'subject_campaigns',
    value,
  }));

export const selectSendingIps = (data = {}) => {
  const sendingIps = data['sending-ips'] || [];
  return sendingIps.map(value => ({
    type: 'sending_ips',
    value,
  }));
};

export const selectIpPools = (data = {}) => {
  const ipPools = data['ip-pools'] || [];
  return ipPools.map(value => ({
    type: 'ip_pools',
    value,
  }));
};

export const selectTemplates = ({ templates = [] } = {}) =>
  templates.map(value => ({
    type: 'templates',
    value,
  }));
