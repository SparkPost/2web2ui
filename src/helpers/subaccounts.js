import _ from 'lodash';

export function setSubaccountQuery(id) {
  return _.isNil(id) ? '' : `?subaccount=${id}`;
}

export function setSubaccountHeader(subaccount = null) {
  const headers = {};

  if (subaccount !== null) {
    headers['x-msys-subaccount'] = _.get(subaccount, 'id', subaccount);
  }

  return headers;
}
