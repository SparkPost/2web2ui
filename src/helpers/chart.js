import moment from 'moment';
import _ from 'lodash';
import { getPrecisionType } from './metrics';
import { roundToPlaces } from 'src/helpers/units';

function getDayLines(data, precision = 'day') {
  if (getPrecisionType(precision) !== 'hour') {
    return [];
  }

  const lastIndex = data.length - 1;
  return data.filter(({ ts }, i) => {
    if (i === 0 || i === lastIndex) {
      return false;
    }
    if (moment(ts).hours() !== 0) {
      return false;
    }
    return true;
  });
}

const getTimeTickFormatter = _.memoize((precisionType, { includeTimezone } = {}) => {
  const formatFunction = includeTimezone ? moment.parseZone : moment;
  const format = tickFormat => tick => formatFunction(tick).format(tickFormat);

  switch (precisionType) {
    case 'hour':
      return format('h:mma');

    case 'day':
    case 'week':
      return format('MMM Do');

    case 'month':
      return format('MMM');

    default:
      return format('MMM Do');
  }
});

const getTooltipLabelFormatter = _.memoize((precisionType, { includeTimezone } = {}) => {
  const formatFunction = includeTimezone ? moment.parseZone : moment;
  const format = labelFormat => label => formatFunction(label).format(labelFormat);
  switch (precisionType) {
    case 'hour':
      return format('MMM Do [at] LT');

    case 'day':
      return format('MMMM Do');

    case 'month':
      return format('MMMM YYYY');

    default:
      return format('MMMM Do');
  }
});

const getWeekPrecisionFormatter = to => {
  return label => {
    const format = momentTime => momentTime.format('MMM Do');
    const startDate = format(moment.parseZone(label));
    //If it's the last date, make the end date the to date; else, make the to date the next Saturday
    if (moment.parseZone(label).isSame(moment.parseZone(to).weekday(0), 'day')) {
      return `${startDate} - ${format(moment.parseZone(to))}`;
    }

    return `${startDate} - ${format(moment.parseZone(label).weekday(6))}`;
  };
};

function getLineChartFormatters(precision, to = moment(), { includeTimezone } = {}) {
  const formatters = {};
  const precisionType = getPrecisionType(precision);

  formatters.xTickFormatter = getTimeTickFormatter(precisionType, { includeTimezone });
  formatters.tooltipLabelFormatter =
    precisionType === 'week' //Can't put in case;switch in getToolTipLabelFormatter b/c memoization
      ? getWeekPrecisionFormatter(to)
      : getTooltipLabelFormatter(precisionType, { includeTimezone });

  return formatters;
}

function cumulativeSum({ data, key }) {
  const results = data.reduce((cumulativeDataSet, data) => {
    const cumulativeData = { ...data };

    if (cumulativeDataSet.length > 0 && data[key]) {
      cumulativeData[key] = data[key] + cumulativeDataSet[cumulativeDataSet.length - 1][key];
    }

    cumulativeDataSet = [...cumulativeDataSet, cumulativeData];

    return cumulativeDataSet;
  }, []);

  return results;
}

const formatYAxisPercent = _.memoize(v => `${roundToPlaces(v, v < 1 ? 3 : 1)}%`);

export {
  cumulativeSum,
  getDayLines,
  getTimeTickFormatter,
  getTooltipLabelFormatter,
  getWeekPrecisionFormatter,
  getLineChartFormatters,
  formatYAxisPercent,
};
