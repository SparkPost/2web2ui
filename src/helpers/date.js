import moment from 'moment';
import config from 'src/config';
import { roundBoundaries } from './metrics';
import { FORMATS } from 'src/constants';
import { format, utcToZonedTime } from 'date-fns-tz';

export const relativeDateOptions = [
  { value: 'hour', label: 'Last Hour' },
  { value: 'day', label: 'Last 24 Hours' },
  { value: '7days', label: 'Last 7 Days' },
  { value: '10days', label: 'Last 10 Days' },
  { value: '14days', label: 'Last 14 Days' },
  { value: '30days', label: 'Last 30 Days' },
  { value: '90days', label: 'Last 90 Days' },
  { value: 'custom', label: 'Custom' },
];

export const relativeDateOptionsIndexed = relativeDateOptions.reduce((result, { value, label }) => {
  result[value] = label;
  return result;
}, {});

export const getRelativeDateOptions = ranges =>
  relativeDateOptions.filter(item => ranges.includes(item.value));

/**
 * Takes a date string and returns the end of that day (11:59PM)
 *
 * If preventFuture is true and the given day IS the current day,
 * it returns the current time, i.e. the closest to the end
 * of the day without going into the future.
 *
 * @param {String} date - date string to base date on
 * @return {Date}
 */
export function getEndOfDay(date, { preventFuture } = {}) {
  const now = moment();
  const end = new Date(date);

  if (preventFuture && now.isSame(end, 'day')) {
    return now.toDate();
  }

  end.setHours(23, 59, 59, 999);
  return end;
}

export function getStartOfDay(date) {
  const start = new Date(date);
  start.setHours(0, 0, 0, 0);
  return start;
}

export function getNextHour(date) {
  const roundedDate = new Date(date);
  const now = new Date(Date.now());
  roundedDate.setHours(now.getHours() + 1, 0, 0, 0);
  return roundedDate;
}

export function isSameDate(a, b) {
  return a instanceof Date && b instanceof Date && a.getTime() === b.getTime();
}

/**
 * Use native JS methods to get user's local
 * timezone, if those methods are present
 *
 * Defaults to UTC otherwise
 */
export function getLocalTimezone() {
  // eslint-disable-next-line new-cap
  if (typeof Intl === 'undefined' || typeof Intl.DateTimeFormat !== 'function') {
    return 'UTC';
  }

  // eslint-disable-next-line new-cap
  const format = Intl.DateTimeFormat();

  if (typeof format.resolvedOptions !== 'function') {
    return 'UTC';
  }

  return format.resolvedOptions().timeZone;
}

export function getRelativeDates(
  range,
  { now = moment().toDate(), roundToPrecision = true, precision } = {},
) {
  let preciseFrom;

  switch (range) {
    case 'hour':
      preciseFrom = moment(now)
        .subtract(1, 'hour')
        .toDate();
      break;

    case 'day':
      preciseFrom = moment(now)
        .subtract(1, 'day')
        .toDate();
      break;

    case '7days':
    case '10days':
    case '14days':
    case '30days':
    case '90days': {
      const numberOfDays = parseInt(range.replace('days', ''));
      preciseFrom = moment(now)
        .subtract(numberOfDays, 'day')
        .toDate();
      break;
    }
    case 'custom':
      return { relativeRange: range };

    default: {
      if (range) {
        throw new Error(
          `Tried to calculate a relative date range with an invalid range value: ${range}`,
        );
      }
      return {};
    }
  }

  if (roundToPrecision) {
    const { to, from } = roundBoundaries({ from: preciseFrom, to: now, precision });
    return { to: to.toDate(), from: from.toDate(), relativeRange: range };
  }

  return { to: now, from: preciseFrom, relativeRange: range };
}

export function getDuration(dates, unit = 'hours') {
  if (!dates) return;

  const { from, to } = dates;

  return moment(to).diff(moment(from), unit);
}

export function formatToTimezone(date, dateFormat, timeZone = getLocalTimezone()) {
  if (!date) {
    throw new Error('Invalid date passed in');
  } else if (!dateFormat) {
    throw new Error('No date formatter passed in');
  }
  return format(utcToZonedTime(new Date(date), timeZone), dateFormat, { timeZone });
}

export function formatDate(date, FORMAT = config.dateFormat) {
  return moment(date).format(FORMAT);
}

export function formatTime(time, FORMAT = config.timeFormat) {
  return moment(time).format(FORMAT);
}

export function formatDateTime(datetime, FORMAT) {
  return FORMAT
    ? moment(datetime).format(FORMAT)
    : `${formatDate(datetime)}, ${formatTime(datetime)}`;
}

export function formatDateTimeWithoutYear(datetime) {
  return formatDateTime(datetime, `${config.dateFormatWithoutYear}, ${config.timeFormat}`);
}

export function getFormattedDateRangeForAggregateData(from, to) {
  return `${formatDate(from, 'MMM Do')} - ${formatDate(to, 'MMM Do, YYYY')}`;
}
// format as ISO 8601 timestamp to match SP API
export const formatApiTimestamp = time => moment.utc(time).format();
export const formatApiDate = date => moment.utc(date).format(FORMATS.MOMENT.SHORT_DATE);
export const formatInputDate = date => moment(date).format(FORMATS.MOMENT.SHORT_DATE);
export const formatInputTime = time => moment(time).format(FORMATS.MOMENT.TIME);
export const parseDate = str => moment(str, FORMATS.MOMENT.INPUT_DATES, true);
export const parseTime = str => moment(str, FORMATS.MOMENT.INPUT_TIMES, true);
export const parseDatetime = (...args) =>
  moment(args.join(' '), FORMATS.MOMENT.INPUT_DATETIMES, true);
export const parseDateTimeTz = (timezone, ...args) =>
  moment.tz(args.join(' '), FORMATS.MOMENT.INPUT_DATETIMES, true, timezone);
export const fillByDate = ({ dataSet, fill = {}, from, to } = {}) => {
  const orderedData = dataSet.sort((a, b) => new Date(a.date) - new Date(b.date));
  let filledDataSet = [];

  for (let time = moment(from), index = 0; time.isBefore(moment(to)); time.add(1, 'day')) {
    const data = orderedData[index];
    const fillData = { ...fill, date: formatInputDate(time) };

    if (!data || data.date !== fillData.date) {
      filledDataSet = [...filledDataSet, fillData];
    } else {
      filledDataSet = [...filledDataSet, data];
      index++;
    }
  }

  return filledDataSet;
};

/**
 * Generates 3 dates based off provided dates
 * Returns first date, middle date, and end
 */
export function getDateTicks({ to, from }) {
  const diff = moment(to).diff(from, 'days');
  const middle = moment(from).add(diff / 2, 'days');

  return [formatInputDate(from), formatInputDate(middle), formatInputDate(to)];
}

// see, https://momentjs.com/docs/#/parsing/unix-timestamp-milliseconds/
export const toMilliseconds = seconds => seconds * 1000;
