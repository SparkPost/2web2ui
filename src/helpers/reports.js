import _ from 'lodash';
import qs from 'qs';
import { getRelativeDates, relativeDateOptions } from 'src/helpers/date';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import { REPORT_BUILDER_FILTER_KEY_MAP } from 'src/constants';
import { list as METRICS_LIST } from 'src/config/metrics';
export function dedupeFilters(filters) {
  return _.uniqBy(filters, stringifyTypeaheadfilter);
}

/**
 * Parses search string
 * @param  {string} search - location.search
 * @returns {Object}
 * {
 *   filters: array of advanced filters
 *   comparisons: array of filters that are being compared
 *   range: specific range for dates, used in place of specific dates
 *   from: javascript date representing beginning of range
 *   to: javascript date representing beginning of range
 *   timezone: string representing timezone
 *   precision: string representing grouping of data based on a time intervals (minute, hour, day, etc.)
 *  report: string representing ID of an active report
 * }
 */
export function parseSearch(search) {
  if (!search) {
    return {};
  }

  const {
    from,
    to,
    range,
    metrics,
    timezone,
    precision,
    filters = [],
    query_filters,
    comparisons,
    report,
    industryBenchmarkMetric,
    industryBenchmarkFilters,
  } = qs.parse(search, { ignoreQueryPrefix: true, arrayLimit: METRICS_LIST.length });
  let ret = {};

  if (report) {
    ret.report = report;
  }

  if (query_filters) {
    try {
      ret.filters = JSON.parse(decodeURI(query_filters));
    } catch {
      ret.filters = [];
    }
  } else {
    const filtersList = (typeof filters === 'string' ? [filters] : filters).map(filter => {
      const parts = filter.split(':');
      const type = parts.shift();
      let value;
      let id;

      // Subaccount filters include 3 parts
      // 'Subaccount:1234 (ID 554):554' -> { type: 'Subaccount', value: '1234 (ID 554)', id: '554' }
      if (type === 'Subaccount') {
        id = parts.pop();
        value = parts.join(':');
      } else {
        value = parts.join(':');
      }
      return { value, type, id };
    });

    ret.filters = mapFiltersToComparators(filtersList);
  }

  if (comparisons) {
    ret.comparisons = comparisons;
  }

  if (metrics) {
    ret.metrics = typeof metrics === 'string' ? [metrics] : metrics;
  }

  const fromTime = new Date(from);
  const toTime = new Date(to);

  if (from && !isNaN(fromTime)) {
    ret.from = fromTime;
  }

  if (to && !isNaN(toTime)) {
    ret.to = toTime;
  }

  if (range) {
    const invalidRange = !_.find(relativeDateOptions, ['value', range]);
    const effectiveRange = invalidRange ? 'day' : range;
    ret = { ...ret, ...getRelativeDates(effectiveRange) };
  }

  if (precision) {
    ret.precision = precision;
  }

  if (timezone) {
    ret.timezone = timezone;
  }

  if (industryBenchmarkMetric) {
    ret.industryBenchmarkMetric = industryBenchmarkMetric;
    if (industryBenchmarkFilters) {
      try {
        ret.industryBenchmarkFilters = JSON.parse(decodeURI(industryBenchmarkFilters));
      } catch {
        ret.industryBenchmarkFilters = { industryCategory: 'all', mailboxProvider: 'all' };
      }
    }
  }

  // filters are used in pages to dispatch updates to Redux store
  return ret;
}

export function mapFiltersToComparators(filters) {
  const mappedFilters = filters.reduce((acc, { value, type, id }) => {
    const filterKey = REPORT_BUILDER_FILTER_KEY_MAP[type];
    acc[filterKey] = acc[filterKey] || { eq: [] };
    acc[filterKey].eq.push(filterKey === 'subaccounts' ? id : value);
    return acc;
  }, {});

  if (_.isEmpty(mappedFilters)) return [];

  return [{ AND: mappedFilters }];
}
