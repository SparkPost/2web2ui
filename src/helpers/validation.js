import _ from 'lodash';
import { formatBytes } from 'src/helpers/units';
import { getDuration } from 'src/helpers/date';
import { toSentence } from 'src/helpers/array';
import { multilineStringToArray } from 'src/helpers/string';
import { isEmailAddress, isEmailLocalPart, isRecipientEmailAddress } from 'src/helpers/email';
import { domainRegex, slugRegex } from './regex';
import isURL from 'validator/lib/isURL';
import Payment from 'payment';
import moment from 'moment';

export const MAXIMUM_EMAILS_ALLOWED = 10;

// validation gate to only check valiation if string is present
export const ifStringPresent = validator => (str = '') =>
  str.trim() === '' ? undefined : validator(str);

export function required(value) {
  return value ? undefined : 'Required';
}

export function email(value) {
  return isEmailAddress(value) ? undefined : 'Invalid Email';
}

export function emails(str) {
  const values = multilineStringToArray(str);
  return values.length && values.every(isEmailAddress)
    ? undefined
    : 'Must be a comma separated list of valid email addresses';
}

/**
 * Determines if a list of comma delimited strings has a length no larger than the max value
 *
 * @param {String} str comma delimited string
 * @param {Number} max maximum length allowed for validation
 * @returns {(String|undefined)} undefined if the length is less than max, otherwise returns an error string
 */
export const maxItems = max => str => {
  const values = multilineStringToArray(str);
  return values.length > max ? `Must contain no more than ${max} items` : undefined;
};

export const noDuplicateItems = str => {
  const values = multilineStringToArray(str);
  const dedupedValues = [...new Set(values)];
  return values.length !== dedupedValues.length ? "Can't contain duplicate items" : undefined;
};

export function emailLocal(value) {
  return isEmailLocalPart(value) ? undefined : 'Invalid Email';
}

export function recipientEmail(value) {
  return isRecipientEmailAddress(value) ? undefined : 'Invalid Email';
}

export function eventsQuery(filter) {
  if (filter.value && !filter.key) {
    return 'Select a Filter';
  }
  return undefined;
}

export function domain(value) {
  return domainRegex.test(value) ? undefined : 'Invalid Domain';
}

export function slug(value) {
  return slugRegex.test(value)
    ? undefined
    : 'Must contain only lowercase letters, numbers, hyphens, and underscores';
}

export function abTestDefaultTemplate(value, formValues, props) {
  // If a matching template ID is found in the list, then the user's entry is valid
  if (props.templates.some(template => template.id === value.id)) {
    return undefined;
  }

  if (formValues.subaccount) {
    return 'Template not available to selected subaccount';
  }

  return 'Template not available to primary account';
}

export function abTestDuration(value, formValues) {
  const testDuration = getDuration(value) + (parseInt(formValues.engagement_timeout, 10) || 24);
  return testDuration > 720
    ? 'Test duration + engagement timeout must be 30 days or less'
    : undefined;
}

export function abTestDistribution(value, formValues) {
  const { default_template, audience_selection, variants } = formValues;

  // only validate if audience selection is percent and all fields are filled
  if (audience_selection === 'percent') {
    let missingValues;

    _.forEach([default_template, ...variants], (variant = {}) => {
      if (!variant.percent) {
        missingValues = true;
      }
    });

    if (missingValues) {
      return;
    }

    const total = _.reduce(
      variants,
      (sum, variant) => sum + parseFloat(variant.percent),
      parseFloat(default_template.percent),
    );
    return total === 100 ? undefined : `Total distribution must equal 100%. Current: ${total}%`;
  }
}

export function startTimeBeforeEndTime(value) {
  return getDuration(value, 'minutes') > 0 ? undefined : 'Start date must be before end date';
}

export function hasNumber(value) {
  return /\d+/.test(value) ? undefined : 'Must have at least 1 number';
}

export function specialCharacter(value) {
  return /[\W_]+/.test(value) ? undefined : 'Must have at least 1 special character (list)';
}

export function hasLetter(value) {
  return /[A-Za-z]+/.test(value) ? undefined : 'Must have at least 1 letter';
}

export function endsWithWhitespace(value) {
  return /[\S]$/.test(value) ? undefined : "Can't end in a whitespace character";
}

export function nonEmptyFile(file) {
  return !file || file.size > 0 ? undefined : 'File must be non-empty';
}

export const fileExtension = _.memoize(function fileExtension(...extensions) {
  const formattedExtensions = extensions.map(extension => `.${extension.toLowerCase()}`);

  return file => {
    if (!file) {
      return;
    }
    const hasValidExtension = formattedExtensions.some(extension =>
      file.name.toLowerCase().endsWith(extension),
    );

    return hasValidExtension
      ? undefined
      : `Must be a ${toSentence(formattedExtensions, 'or')} file`;
  };
});

export const maxLength = _.memoize(function maxLength(length) {
  return value =>
    value && value.trim().length > length ? `Must be ${length} characters or less` : undefined;
});

export const minLength = _.memoize(function minLength(length) {
  return value =>
    typeof value !== 'undefined' && value.trim().length < length
      ? `Must be at least ${length} characters`
      : undefined;
});

export const integer = value => (/^-?[0-9]+$/.test(value) ? undefined : 'Integers only please');

export const minNumber = _.memoize(function minNumber(min) {
  return value => (value < min ? `Must be at least ${min}` : undefined);
});

export const maxNumber = _.memoize(function maxNumber(max) {
  return value => (value > max ? `Must be less than ${max}` : undefined);
});

export const numberBetween = _.memoize(function numberBetween(min, max) {
  return value => (value > min && value < max ? undefined : `Must be between ${min} and ${max}`);
});

export const numberBetweenInclusive = _.memoize(function numberBetween(min, max) {
  return value => (value >= min && value <= max ? undefined : `Must be between ${min} and ${max}`);
});

export const maxFileSize = _.memoize(function maxFilesSize(maxSize) {
  return file => {
    if (!file) {
      return undefined;
    }
    return file.size < maxSize ? undefined : `Please keep file size under ${formatBytes(maxSize)}`;
  };
});

export function url(value) {
  return isURL(value) ? undefined : 'Must be a valid URL';
}

export const cardExpiry = value =>
  Payment.fns.validateCardExpiry(value) ? undefined : 'Please choose a valid expiration date';

export const json = value => {
  try {
    JSON.parse(value);
  } catch (e) {
    return 'Must be valid JSON';
  }
};

// Date validator for the DatePicker component
export const minDays = _.memoize(function minDays(min) {
  return function(dates) {
    // This checks if the dates meet the minimum, and does not factor in the time of those dates
    // Min is subtracted by 1 to account for up to one 24 hour period excluded from the calculation
    return Math.abs(moment(dates.from).diff(moment(dates.to), 'days')) < min - 1
      ? `Select a range of at least ${min} days`
      : undefined;
  };
});
