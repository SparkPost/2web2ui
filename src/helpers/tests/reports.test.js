import * as reports from '../reports';
import * as dateHelpers from 'src/helpers/date';
import { list as METRICS_LIST } from 'src/config/metrics';
import cases from 'jest-in-case';

jest.mock('src/helpers/date');
jest.mock('src/helpers/string', () => ({
  stringifyTypeaheadfilter: jest.fn(filter => filter.id),
}));

describe('report helpers', () => {
  describe('parseSearch', () => {
    const filters = 'filters=Domain:test.com&filters=Subaccount:test:123';
    const metrics = 'metrics=count-something';

    beforeEach(() => {
      dateHelpers.getRelativeDates = jest.fn(() => ({}));
      dateHelpers.relativeDateOptions = [{ value: '7days', label: 'Last 7 Days' }];
    });

    it('should parse search with missing range', () => {
      const date = 'from=2017-11-03T14:43:00Z&to=2017-11-04T14:43:00Z';
      const search = `?${filters}&${date}&${metrics}`;

      expect(reports.parseSearch(search)).toMatchSnapshot();
    });

    it('should parse search with no empty value', () => {
      expect(reports.parseSearch('').options).toEqual({});
    });

    it('should parse search with a colon in a subaccount name', () => {
      const search = '?filters=Subaccount%3ASubaccount%20with%20a%20bad%3A%20symbol%3A100';
      expect(reports.parseSearch(search).filters).toEqual([
        {
          id: '100',
          type: 'Subaccount',
          value: 'Subaccount with a bad: symbol',
        },
      ]);
    });

    it('uses to relative range with valid from & to', () => {
      const search = `?${filters}&range=7days&from=2019-02-09T02%3A00%3A00Z&to=2019-02-16T02%3A59%3A59Z&${metrics}`;
      reports.parseSearch(search);
      expect(dateHelpers.getRelativeDates).toHaveBeenCalledWith('7days');
    });

    it('uses to default to one day with an invalid range', () => {
      const search = `?${filters}&range=1millionyears&from=2019-02-09T02%3A00%3A00Z&to=2019-02-16T02%3A59%3A59Z&${metrics}`;
      reports.parseSearch(search);
      expect(dateHelpers.getRelativeDates).toHaveBeenCalledWith('day');
    });

    it('defaults to relative range for invalid custom from or to', () => {
      reports.parseSearch(`?${filters}&range=week&from=aaa&to=2019-02-16T02%3A59%3A59Z&${metrics}`);
      expect(dateHelpers.getRelativeDates).toHaveBeenCalledWith('day');

      dateHelpers.getRelativeDates.mockClear();
      reports.parseSearch(
        `?${filters}&range=week&from=2019-02-09T02%3A00%3A00Z&to=2019-02-16T02%3A59%3A59Zzzzzz&${metrics}`,
      );
      expect(dateHelpers.getRelativeDates).toHaveBeenCalledWith('day');
    });

    it('parses long list of metrics to Array correctly', () => {
      const metricsBeer = new Array(21).fill(0).reduce((accumulator, _value, index) => {
        return `${accumulator}&metrics[${index}]=${METRICS_LIST[index].key}`;
      }, '');
      //Because it's over 21
      //Formatted like metrics[0]=foo&metrics[1]=bar

      const { metrics } = reports.parseSearchNew(`?range=week${metricsBeer}`);
      expect(Array.isArray(metrics)).toEqual(true);
    });

    cases(
      'handles invalid datetimes with custom format',
      opts => {
        expect(reports.parseSearch(opts.search).options).toEqual(opts.match);
      },
      [
        {
          name: 'valid from',
          search: 'from=2017-11-03T14:43:00Z',
          match: { from: new Date('2017-11-03T14:43:00.000Z') },
        },
        { name: 'invalid from', search: 'from=2017-11-03T14:43:00Zz', match: {} },
        {
          name: 'invalid from with custom range',
          search: 'from=2017-11-03T14:43:00Zz&range=custom',
          match: {},
        },
        {
          name: 'invalid from valid to with custom range',
          search: 'from=2017-11-03T14:43:00Zk&range=custom&to=2017-11-04T14:43:00Z',
          match: { to: new Date('2017-11-04T14:43:00Z') },
        },
        {
          name: 'valid from invalid to with custom range',
          search: 'from=2017-11-03T14:43:00Z&range=custom&to=2017-11-04T14:43:00Zk',
          match: { from: new Date('2017-11-03T14:43:00Z') },
        },
        {
          name: 'invalid from invalid to with custom range',
          search: 'from=2017-11-03T14:43:00Zk&range=custom&to=2017-11-04T14:43:00Zk',
          match: {},
        },
        {
          name: 'invalid from invalid to with day range',
          search: 'from=2017-11-03T14:43:00Zk&range=day&to=2017-11-04T14:43:00Zk',
          match: {},
        },
      ],
    );
  });

  describe('dedupeFilters', () => {
    it('returns deduped filters', () => {
      const filters = [
        { type: 'Subaccount', value: 'created (ID 698)', id: 698 },
        { type: 'Subaccount', value: 'test create no key (ID 656)', id: 656 },
        { type: 'Subaccount', value: 'created (ID 698)', id: 698 },
        { type: 'not-subaccount', value: 'some random value', id: 550 },
      ];
      expect(reports.dedupeFilters(filters)).toMatchSnapshot();
    });
  });

  describe('mapFiltersToComparators', () => {
    it('returns the new structure for filters', () => {
      const filters = [
        { type: 'Subaccount', value: 'created (ID 698)', id: 698 },
        { type: 'Subaccount', value: 'test create no key (ID 656)', id: 656 },
        { type: 'Template', value: 'oiwjefoijwef' },
        { type: 'Recipient Domain', value: 'some random value' },
      ];
      expect(reports.mapFiltersToComparators(filters)).toMatchSnapshot();
    });

    it('returns an empty array when no mapped filters are returned', () => {
      const filters = [];
      expect(reports.mapFiltersToComparators(filters)).toMatchSnapshot();
    });
  });
});
