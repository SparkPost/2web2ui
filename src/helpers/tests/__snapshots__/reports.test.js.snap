// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report helpers dedupeFilters returns deduped filters 1`] = `
Array [
  Object {
    "id": 698,
    "type": "Subaccount",
    "value": "created (ID 698)",
  },
  Object {
    "id": 656,
    "type": "Subaccount",
    "value": "test create no key (ID 656)",
  },
  Object {
    "id": 550,
    "type": "not-subaccount",
    "value": "some random value",
  },
]
`;

exports[`report helpers mapFiltersToComparators returns an empty array when no mapped filters are returned 1`] = `Array []`;

exports[`report helpers mapFiltersToComparators returns the new structure for filters 1`] = `
Array [
  Object {
    "AND": Object {
      "domains": Object {
        "eq": Array [
          "some random value",
        ],
      },
      "subaccounts": Object {
        "eq": Array [
          698,
          656,
        ],
      },
      "templates": Object {
        "eq": Array [
          "oiwjefoijwef",
        ],
      },
    },
  },
]
`;

exports[`report helpers parseSearch should parse search with missing range 1`] = `
Object {
  "filters": Array [
    Object {
      "id": undefined,
      "type": "Domain",
      "value": "test.com",
    },
    Object {
      "id": "123",
      "type": "Subaccount",
      "value": "test",
    },
  ],
  "options": Object {
    "from": 2017-11-03T14:43:00.000Z,
    "metrics": Array [
      "count-something",
    ],
    "to": 2017-11-04T14:43:00.000Z,
  },
  "report": undefined,
}
`;
